From eb38b5e4bf08622f9b69adce2340a3395057b6c2 Mon Sep 17 00:00:00 2001
From: Andrew Morgan <andrew@amorgan.xyz>
Date: Wed, 4 Dec 2024 18:33:47 +0000
Subject: [PATCH] pass an instance, not a type, to resolveHostName

resolveHostName wanted an instance of a type. We also had to switch from provider to
implementer, as the latter implies that any instances implement the given interface.

Fixed the following error:

synapse/http/client.py:243: error: Argument 1 to "resolveHostName" of "IHostnameResolver" has incompatible type "type[EndpointReceiver]"; expected "IResolutionReceiver"  [arg-type]
---
 synapse/http/client.py | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/synapse/http/client.py b/synapse/http/client.py
index 85923d956b..559b1febf0 100644
--- a/synapse/http/client.py
+++ b/synapse/http/client.py
@@ -41,7 +41,7 @@ from canonicaljson import encode_canonical_json
 from netaddr import AddrFormatError, IPAddress, IPSet
 from prometheus_client import Counter
 from typing_extensions import Protocol
-from zope.interface import implementer, provider
+from zope.interface import implementer
 
 from OpenSSL import SSL
 from OpenSSL.SSL import VERIFY_NONE
@@ -225,7 +225,7 @@ class _IPBlockingResolver:
                     recv.addressResolved(address)
             recv.resolutionComplete()
 
-        @provider(IResolutionReceiver)
+        @implementer(IResolutionReceiver)
         class EndpointReceiver:
             @staticmethod
             def resolutionBegan(resolutionInProgress: IHostResolution) -> None:
@@ -239,8 +239,9 @@ class _IPBlockingResolver:
             def resolutionComplete() -> None:
                 _callback()
 
+        endpoint_receiver_wrapper = EndpointReceiver()
         self._reactor.nameResolver.resolveHostName(
-            EndpointReceiver, hostname, portNumber=portNumber
+            endpoint_receiver_wrapper, hostname, portNumber=portNumber
         )
 
         return recv
-- 
2.46.2

